### PR TITLE
Create a secp256r1 resolution method

### PR DESCRIPTION
**Pull request by bshambaugh**
I used JsonWebKey2020 which uses publicJwk to map to an object? This is the first use of publicJwk. secp256k1 and ed25519 uses publicKeyBase58 which uses a string.

### CODE REVIEW

File: packages/key-did-resolver/src/secp256r1.ts
Comment (oed) [inline packages/key-did-resolver/src/secp256r1.ts:16]: What's the reasoning behind using `JsonWebKey2020` instead of `publicKeyBase58`?

File: packages/key-did-resolver/src/__tests__/secp256r1.test.ts
Comment (oed) [inline packages/key-did-resolver/src/__tests__/secp256r1.test.ts:8]: Afaik the Key DIDs should only contain the X value (see secp256k1 and ed25519 examples).

File: packages/key-did-resolver/src/secp256r1.ts
Comment (oed) [inline packages/key-did-resolver/src/secp256r1.ts:37]: This would break if the public key in the did string only contains the x value.

**Review by oed — COMMENTED** [review]
Not sure this is completely right. From the other examples we always just include the X value in the public key part of the did string. Seems like you are including both X and Y.

File: package.json
Comment (bshambaugh) [reply package.json:42]: Okay, fair enough. I needed it install this dependency to get js-ceramic's tests for this package to work on my own machine.

Comment (bshambaugh) [reply package.json:42]: I can remove it from the PR.

File: packages/key-did-resolver/src/secp256r1.ts
Comment (bshambaugh) [reply packages/key-did-resolver/src/secp256r1.ts:16]: I will ask.

Comment (bshambaugh) [reply packages/key-did-resolver/src/secp256r1.ts:16]: See https://lists.w3.org/Archives/Public/public-credentials/2021Apr/0139.html .

Comment (oed) [reply packages/key-did-resolver/src/secp256r1.ts:16]: Ok, thanks for clarifying. I think this makes sense then.

File: packages/key-did-resolver/src/__tests__/secp256r1.test.ts
Comment (oed) [reply packages/key-did-resolver/src/__tests__/secp256r1.test.ts:8]: does the transmute-industries/did-key.js also use this format?

Comment (oed) [reply packages/key-did-resolver/src/__tests__/secp256r1.test.ts:8]: If so this should be alright.

File: package.json
Comment (oed) [reply package.json:42]: Yes please :)

File: packages/key-did-resolver/src/__tests__/secp256r1.test.ts
Comment (bshambaugh) [reply packages/key-did-resolver/src/__tests__/secp256r1.test.ts:8]: Transmute industries may be inconsistent with the present state. I believe Orie is updating a part before becoming consistent with the did method key spec and this issue.
https://github.com/w3c-ccg/did-method-key/issues/29#issuecomment-786028270

However, the did key is the same length as in: https://w3c-ccg.github.io/did-method-key/#p-256 .

Comment (bshambaugh) [reply packages/key-did-resolver/src/__tests__/secp256r1.test.ts:8]: edit: Orie is waiting on bls12381 which is not working with node 15. https://github.com/transmute-industries/did-key.js/pull/78#issuecomment-806200291

Comment (bshambaugh) [reply packages/key-did-resolver/src/__tests__/secp256r1.test.ts:8]: Fwiw, I am going back to see how I created the did:key. The code works with the did:key examples from the ccg spec and the ones that I created on my own.

Comment (bshambaugh) [reply packages/key-did-resolver/src/__tests__/secp256r1.test.ts:8]: This is my code to create the did:key :: https://github.com/bshambaugh/key-did-provider-secp256r1

Comment (bshambaugh) [reply packages/key-did-resolver/src/__tests__/secp256r1.test.ts:8]: Yes. The hex that I used to create the key is 64 bytes. It is a hex string that has a uint8array of bytes that looks like this: https://www.youtube.com/watch?v=_MjjF211BM8&t=263s .

Looking at the code and output suggests that 64 bytes is the full X,Y key. 
https://gist.github.com/bshambaugh/36aa03e0d79b630678c558f32c6ed936

File: packages/key-did-resolver/src/secp256r1.ts
Comment (bshambaugh) [reply packages/key-did-resolver/src/secp256r1.ts:37]: I believe that I have the X,Y pair. My did:key is of the same length as the did:key spec.

File: packages/key-did-resolver/src/__tests__/secp256r1.test.ts
Comment (bshambaugh) [reply packages/key-did-resolver/src/__tests__/secp256r1.test.ts:8]: I can ask if the compressed public key would be better for did:key

Comment (oed) [reply packages/key-did-resolver/src/__tests__/secp256r1.test.ts:8]: Yeah, I'm mostly just surprised why `secp256k1` is compressed while `r1` is not. I'm fine with leaving it as is though since it seems like that's what the did:key spec says.

Comment (bshambaugh) [reply packages/key-did-resolver/src/__tests__/secp256r1.test.ts:8]: Kyle Hartog believed the spec was the best thing to look at. He also believed that the work was stalled. 
https://lists.w3.org/Archives/Public/public-credentials/2021Apr/0171.html
I am currently investigating how difficult it would be to modify the function to convert the compressed form into the uncompressed form for resolving the did doc.

Comment (bshambaugh) [reply packages/key-did-resolver/src/__tests__/secp256r1.test.ts:8]: There is some conversation going on over here about compressed vs. uncompressed did:keys. I am linking to it to keep track of it all: https://github.com/w3c-ccg/did-method-key/issues/32

Comment (oed) [reply packages/key-did-resolver/src/__tests__/secp256r1.test.ts:8]: Thanks. Should we wait for that to be resolved before we merge this?

Comment (bshambaugh) [reply packages/key-did-resolver/src/__tests__/secp256r1.test.ts:8]: The last link [1] that dlongley linked to in his Post is interesting.

[1] https://stackoverflow.com/questions/17171542/algorithm-for-elliptic-curve-point-compression/30431547#30431547
[2] https://github.com/w3c-ccg/did-method-key/issues/32#issuecomment-827681999

I found out that modPow is used for the decompression. Presently I am trying to perform the same functions by not using the JavaScript Big Integer Library. modPow is for modular exponentiation [4].

[3] https://github.com/peterolson/BigInteger.js
[4] https://en.wikipedia.org/wiki/Modular_exponentiation

BigInt [5] is included natively in JavaScript now, which allows me to emulate some of the functionality of the [1] link.

However, I am struggling to implement modPow. Assuming:

var yBig = xBig.pow(3).sub( xBig.multiply(3) ).add( b ).modPow( pIdent, prime );

is the same as

var yBig = ((xBig**3n - xBig*3n + b)**pIdent) % prime;

gives me an error: "RangeError: Maximum BigInt size exceeded"

To solve this error, I looked for implementations of modPow with JavaScript. I found this one [6] that works with BigInt.

[5] https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt
[6] https://github.com/juanelas/bigint-mod-arith

or something like this [7] if the library is to be avoided

[7] https://gist.github.com/krzkaczor/0bdba0ee9555659ae5fe

Should I include a library, or should I add a function for implementing BigInt to my code?

Comment (oed) [reply packages/key-did-resolver/src/__tests__/secp256r1.test.ts:8]: > Should I include a library, or should I add a function for implementing BigInt to my code?

Which library are you referring to? If it's not too big then I think it's fine to include.

Comment (bshambaugh) [reply packages/key-did-resolver/src/__tests__/secp256r1.test.ts:8]: I have raw (64 byte), compressed (33 byte), and uncompressed (65 byte) keys in my code now. My tests pass, but I do not have full coverage yet.
scr-1620860401

**Comment by oed** [comment]
@bshambaugh please let us know when this is ready for a full review again :)

File: packages/key-did-resolver/src/__tests__/secp256r1.test.ts
Comment (bshambaugh) [reply packages/key-did-resolver/src/__tests__/secp256r1.test.ts:10]: Ahh...line 61 was for the function ECPointDecompress . I added an extra test and it looks to be covered. The conditional if block that line 61 is in only executes if the parity of the compressed key is opposite (of what is should be?).

Comment (bshambaugh) [reply packages/key-did-resolver/src/__tests__/secp256r1.test.ts:10]: scr-1621014068

Now I need to address the pubKeyBytesToXY function. I believe the function will fail if the bytesCount is 65 and the 1st byte is not 03, the bytes count is 33 and the 1st bytes is not 03 or 02, or the bytes count is not 64, 65, or 33. I guess I could have the function return null if this is the case. I am not sure how the did:key document will resolve if the function returns null. I could try to report null in place of the key, or give an error message instead of resolving the document will nulls.

Comment (bshambaugh) [reply packages/key-did-resolver/src/__tests__/secp256r1.test.ts:10]: scr-1621027899

I have full coverage. Unfortunately, I found some tests that I can create that can fail. I have not included them yet since they use functions included to enable the tests, not functions that are in the code. I think that the same bug may be in the code too. I am going to try to see if I can fix the bug in the test function and the code functions.

test that fail:

test('test a hex string shorter than 33 bytes', () => {
   const inputPublicKeyHex = '36f8964623378bdc068d4bce07ed17c8fa486f9ac0c2613ca3c8c306d7bb6'
   const publicKey_u8a = pubKeyHexToUint8Array(inputPublicKeyHex);
   const pubKeyBytesToXY = mapper.pubKeyBytesToXY(publicKey_u8a);
   expect(pubKeyBytesToXY).toEqual(null);
});


test('test a hex string between 64 and 33 bytes', () => {
   const inputPublicKeyHex = '36f8964623378bdc068d4bce07ed17c8fa486f9ac0c2613ca3c8c306d7bb61cd36717b8ac5e4fea8ad23dc8d0783c2318ee4ad7a80db6e0026ad0b072a24f'
   const publicKey_u8a = pubKeyHexToUint8Array(inputPublicKeyHex);
   const pubKeyBytesToXY = mapper.pubKeyBytesToXY(publicKey_u8a);
   expect(pubKeyBytesToXY).toEqual(null);
});

**Comment by bshambaugh** [comment]
> @bshambaugh please let us know when this is ready for a full review again :)

I'm dipping my toes into the case where I have nulls coming into the functions:

https://gist.github.com/bshambaugh/333a4d28e3fdb275ed20422e97030e04

**Comment by oed** [comment]
There's something wrong with this PR theres 269 changed files. Looks like you included build outputs in the src folder somehow. Please fix this and I'll do a final review!

**Comment by bshambaugh** [comment]
scr-1621607332
scr-1621607383

There is not full test coverage for index.ts

**Review by oed — CHANGES_REQUESTED** [review]
This looks great. Just add the missing dependency `bigint-mod-arith` to the package.json. Will merge after this is done.

Tests are passing locally when adding the missing dependency.

**Comment by oed** [comment]
Still seeing issues with this branch. Can you try to do the following:

**Merge develop into this branch:**
```
git merge develop
```

**Make sure build and lint passes:**
```
npm run lint
// and
npm run build
```

That should be it.

**Comment by bshambaugh** [comment]
> Still seeing issues with this branch. Can you try to do the following:
> 
> **Merge develop into this branch:**
> 
> ```
> git merge develop
> ```
> 
> **Make sure build and lint passes:**
> 
> ```
> npm run lint
> // and
> npm run build
> ```
yes, I still need to work through lint and build
> 
> That should be it.

**Comment by bshambaugh** [comment]
In addition to keyToDidDoc which is used by 
https://github.com/ceramicnetwork/js-ceramic/blob/develop/packages/key-did-resolver/src/index.ts , the following functions are exported for testing purposes:

https://github.com/bshambaugh/js-ceramic/blob/develop/packages/key-did-resolver/src/secp256r1.ts#L54
ECPointDecompress

https://github.com/bshambaugh/js-ceramic/blob/develop/packages/key-did-resolver/src/secp256r1.ts#L80
publicKeyToXY

https://github.com/bshambaugh/js-ceramic/blob/develop/packages/key-did-resolver/src/secp256r1.ts#L87
publicKeyHexToUint8ArrayPointPair

https://github.com/bshambaugh/js-ceramic/blob/develop/packages/key-did-resolver/src/secp256r1.ts#L95
publicKeyIntToXY

https://github.com/bshambaugh/js-ceramic/blob/develop/packages/key-did-resolver/src/secp256r1.ts#L102
publicKeyIntToUint8ArrayPointPair

https://github.com/bshambaugh/js-ceramic/blob/develop/packages/key-did-resolver/src/secp256r1.ts#L110
pubKeyBytesToXY

With the exception of pubKeyBytesToXY, all of these functions are known to fail for null and undefined.

**Comment by bshambaugh** [comment]
To my knowledge, everything looks good.

**Review by oed — APPROVED** [review]
Looks great 👍 

Btw, I made a PR to the multicodecs table to clarify that the compressed format should be used: https://github.com/multiformats/multicodec/pull/214

**Comment by oed** [comment]
@bshambaugh looks like the docs are not building. My bad for not mentioning this to you. Could you make another PR to make sure that `npm run docs` passes?

**Comment by bshambaugh** [comment]
key-did-resolver$ npm run docs
npm ERR! Missing script: "docs"
npm ERR! 
npm ERR! To see a list of scripts, run:
npm ERR!   npm run

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/ubuntu/.npm/_logs/2021-05-25T13_23_41_860Z-debug.lo

**Comment by oed** [comment]
@bshambaugh We had to revert the change because the docs script didn't pass. Can you please submit a new PR for this feature which includes working docs? 🙏

**Pull request merged**
